### ISSUE TITLE
feature: more configurable symbol filters

### ISSUE THREAD

**Issue by emmanueltouzery** [stevearc/aerial.nvim#214]
so here's with typescript and typescript react... and using the LSP backend, which is i believe better for that language..
many times in typescript (and javascript), it's better to define functions as constants:
```javascript
const myFunction = () => ...
```
instead of
```javascript
function myFunction() { ...
```

So that means that to actually see these functions, we must configure aerial with filter_kind containing Constant, problem solved, right?
Well, no, because then we get all the constants defined in the source code.. which is not what we want. The workaround I found for that is to filter these to only those that have children. That at least removes all the really simple constants. But it's a heuristic only.

And then.. the LSP will register as functions all the lambdas that we have all over in functions... for instance `x.map(() ->).filter(() -> )` will register two functions. But it registers these either with a symbol name ending with " callback" or with a symbol name of "&lt;function&gt;".

So again we can filter these by hand.

All of this to explain my feature request, which is to enable to configure which symbols to hide through callbacks in the aerial configuration. Example:

```lua
      filter_function = {
        ['_'] = nil,
        typescript = function(symbol)
          local kind = get_symbol_kind_name(symbol.kind)
          if kind == 'Constant' then
            return #symbol.children > 0
          end
          if kind == 'Function' then
            return string.match(symbol.name, " callback$") == nil and symbol.name ~= "<function>"
          end
          return true
        end
      }
```
No filter function in general, this special function for typescript. And this assume that `filter_kind` filtering is applied before this one, but this one could handle both.

With something like that, these kind of advanced (but I think relatively common) use-cases can be covered...

**Comment by stevearc** [comment]
I could potentially build in language-specific logic, or extension points, for the LSP backend, but first I'd like to see if that's really necessary. The treesitter backend already has some hooks for language-specific logic, so my first instinct would be to ask in what way the LSP backend is superior to the treesitter one? It's probably going to be easier to improve the treesitter backend than add a bunch of new logic to the LSP one.

**Comment by emmanueltouzery** [comment]
Ah yes... I didn't check the tree-sitter backend at all, I just assumed "of course", the LSP one would be better.

I checked it now. And I see you already have some custom support for javascript and typescript (note that there is also typescriptreact... about the same as typescript, except it can contain JSX tags.. and also javascriptreact, same thing.. possibly the rules can be reused but I'm not sure).

So I made a small sample file now:
```javascript
function test() {
  const x = 2;
  return x;
}

const f = createFunction(12);

export const f2 = createFunction(15);

const f3 = () => {
  return [1,2,3].map(() => 5);
};

const val = 12;
```

The outcome I'd hope for would be:
test, f, f2, f3

with the default filter_kind, I get for both TS and JS: test, f3

If I expand the filter_kind to:
```lua
        javascript = { "Class", "Constructor", "Enum", "Function", "Interface", "Module", "Method", "Struct", "Variable", "Type" },
        typescript = { "Class", "Constructor", "Enum", "Function", "Interface", "Module", "Method", "Struct", "Variable", "Type" },
```

then I get for TS: test, test.x, f, f2, f3, val, and for JS still test, f3.

If we take this second output for TS, and just filter out the nested variables, meaning test.x, it's more or less right. But then what is the nesting, if it's a class, then presumably we want to keep the functions.. While if it's a function we don't want to :(

I didn't expect you'd be ready to take on that problem in the library, i expected each person would be expected to fix it from themselves in their setup.

**Comment by emmanueltouzery** [comment]
So basically, since it's impossible without a type-checker to prove that these constants are in fact functions...

I'd like to be able to display constants that are not under functions. Meaning that are at the toplevel or under classes. So in the example that I gave, `x` is a false positive.

**Comment by stevearc** [comment]
Yep, I came to the same conclusion. Treesitter has no (simple, easy) way to get the runtime type of a constant, so for what you're asking we would have to use LSP. This will take some time to figure out a good API, and I'm honestly a bit short on free time these days. If you want to draft a PR I'd be happy to do some back-and-forth on the approach, otherwise it might be a while until I get to this.

**Comment by emmanueltouzery** [comment]
You would try to reach the type through the LSP? I was hoping that my workaround idea of checking the context in which a constant is declared would be achievable through tree sitter queries alone --and therefore no API change, just tricky -for me anyway- tree sitter queries.

But then it's really more a workaround...

If checking the type through the LSP, we'd probably do that directly from the lsp plugin and not try to support that with the tree-sitter backend, right?

**Comment by emmanueltouzery** [comment]
Note that I'm not sure whether getting the type from the LSP would be enough to know whether it's really a function, with type aliases, generics and so on. I don't know much about it, but I'm assuming we'd get the type as a string, and I'd be a priori pessimistic amour interpreting it.

**Comment by emmanueltouzery** [comment]
I'm thinking maybe this could be achieved in the javascript/typescript tree-sitter post-processing that you mentioned in #216. Have tree-sitter queries report all the constants, then the post-processing could filter only the suitably top-level ones. I'll give it a shot.

**Comment by emmanueltouzery** [comment]
Fixed by #220
